compiler/src/iree/compiler/Codegen/Common/GPU/GPUDistributionPatterns.cpp

```diff
@@ -20,6 +22,82 @@ using VectorValue = TypedValue<VectorType>;
 
 namespace {
 
+static std::optional<int64_t> findDimInLayout(LayoutAttr layout,
+                                              LayoutDimension dim) {
+  for (PerDimLayoutAttr dimLayout : layout.getLayouts()) {
+    if (std::optional<int64_t> shape = dimLayout.getShape(dim)) {
+      return shape;
+    }
+  }
+  return std::nullopt;
+}
+
+static SmallVector<Value> computeSIMDIndex(const LayoutIterator::State &state,
+                                           LayoutAttr layout,
+                                           ArrayRef<Value> threadGrid,
+                                           RewriterBase &rewriter) {
+  MLIRContext *ctx = layout.getContext();
+  AffineExpr threadX, threadY, threadZ;
+  bindSymbols(ctx, threadX, threadY, threadZ);
+
+  SmallVector<Value> simdIndex;
+  // Calculate the index for each dim separately.
+  for (PerDimLayoutAttr dimLayout : layout.getLayouts()) {
+    AffineExpr offset = getAffineConstantExpr(0, ctx);
+    AffineExpr stride = getAffineConstantExpr(1, ctx);
+    for (const auto &[label, shape] : llvm::reverse(
```

Comment:
```suggestion
    for (auto [label, shape] : llvm::reverse(
```
`const &` doesn't prevent copies or modification here